managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -234,11 +235,49 @@ public MarkDeleteEntry(PositionImpl newPosition, Map<String, Long> properties,
         return lastMarkDeleteEntry != null ? lastMarkDeleteEntry.properties : Collections.emptyMap();
     }
 
+    @Override
+    public void asyncAddEntries(byte[] entry, AsyncCallbacks.AddEntryCallback callback, Object ctx) {
+        cursorLedger.asyncAddEntry(entry, (rc, lh, entryId, addCtx) -> {
+            if (rc == BKException.Code.OK) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Add entry to cursor {} ledger id: {}, entry id: {}", ledger.getName(), name,
+                              lh.getId(), entryId);
+                }
+
+                callback.addComplete(new PositionImpl(lh.getId(), entryId), addCtx);
+            } else {
+                log.warn("");
+                callback.addFailed(createManagedLedgerException(rc), addCtx);
+            }
+        }, ctx);
+    }
+
+    public CompletableFuture<Long> getLastEntryId() {
+        return cursorLedger.tryReadLastAddConfirmedAsync();
+    }
+
+    @Override
+    public void asyncReadEntry(long entryId, ReadEntryCallback callback, Object ctx) {
+        cursorLedger.asyncReadEntries(entryId, entryId, (rc, lh, entries, readCtx) -> {
+            if (rc == BKException.Code.OK) {
+                if (log.isDebugEnabled()) {
+                    log.debug("");
+
+                }
+                LedgerEntry ledgerEntry = entries.nextElement();
```

Comment:
A good practice is to check if entries hasMoreElements before calling nextElement